File: pkg/ui/workspaces/cluster-ui/src/util/appStats/appStats.ts
// Copyright 2021 The Cockroach Authors.
//
// Use of this software is governed by the CockroachDB Software License
// included in the /LICENSE file.

import { cockroach } from "@cockroachlabs/crdb-protobuf-client";
import Long from "long";

import { uniqueLong, unique } from "src/util/arrays";
import { TimestampToNumber, DurationToNumber } from "src/util/convert";
import { FixLong } from "src/util/fixLong";

export type StatementStatistics = cockroach.sql.IStatementStatistics;
export type ExecStats = cockroach.sql.IExecStats;
export type CollectedStatementStatistics =
  cockroach.server.serverpb.StatementsResponse.ICollectedStatementStatistics;

export interface NumericStat {
  mean?: number;
  squared_diffs?: number;
}

export function variance(stat: NumericStat, count: number): number {
  return (stat?.squared_diffs || 0) / (count - 1);
}

export function stdDev(stat: NumericStat, count: number): number {
  return Math.sqrt(variance(stat, count)) || 0;
}

export function stdDevLong(stat: NumericStat, count: number | Long): number {
  return stdDev(stat, Number(FixLong(count)));
}

// aggregateNumericStats computes a new `NumericStat` instance from 2 arguments by using the counts
// to generate new `mean` and `squared_diffs` values.
export function aggregateNumericStats(
  a: NumericStat,
  b: NumericStat,
  countA: number,
  countB: number,
): { mean: number; squared_diffs: number } {
  const total = countA + countB;
  const delta = b?.mean - a?.mean;

  return {
    mean: (a?.mean * countA + b?.mean * countB) / total,
    squared_diffs:
      a?.squared_diffs +
      b?.squared_diffs +
      (delta * delta * countA * countB) / total,
  };
}

export function aggregateLatencyInfo(
  a: StatementStatistics,
  b: StatementStatistics,
): cockroach.sql.ILatencyInfo {
  const min =
    a.latency_info?.min === 0 || a.latency_info?.min > b.latency_info?.min
      ? b.latency_info?.min
      : a.latency_info?.min;
  const max =
    a.latency_info?.max > b.latency_info?.max
      ? a.latency_info?.max
      : b.latency_info?.max;

  let p50 = b.latency_info?.p50;
  let p90 = b.latency_info?.p90;
  let p99 = b.latency_info?.p99;
  // Use the latest value we have that is not zero.
  if (
    b.last_exec_timestamp < a.last_exec_timestamp &&
    b.latency_info?.p50 !== 0
  ) {
    p50 = a.latency_info?.p50;
    p90 = a.latency_info?.p90;
    p99 = a.latency_info?.p99;
  }

  return {
    min,
    max,
    p50,
    p90,
    p99,
  };
}

export function coalesceSensitiveInfo(
  a: cockroach.sql.ISensitiveInfo,
  b: cockroach.sql.ISensitiveInfo,
): cockroach.sql.ISensitiveInfo {
  return {
    last_err: a.last_err || b.last_err,
    most_recent_plan_description:
      a.most_recent_plan_description || b.most_recent_plan_description,
  };
}

export function addMaybeUnsetNumericStat(
  a: NumericStat,
  b: NumericStat,
  countA: number,
  countB: number,
): NumericStat {
  return a && b ? aggregateNumericStats(a, b, countA, countB) : null;
}

export function addExecStats(a: ExecStats, b: ExecStats): ExecStats {
  let countA = FixLong(a.count).toInt();
  const countB = FixLong(b.count).toInt();
  if (countA === 0 && countB === 0) {
    // If both counts are zero, artificially set the one count to one to avoid
    // division by zero when calculating the mean in addNumericStats.
    countA = 1;
  }
  return {
    count: a.count.add(b.count),
    network_bytes: addMaybeUnsetNumericStat(
      a.network_bytes,
      b.network_bytes,
      countA,
      countB,
    ),
    max_mem_usage: addMaybeUnsetNumericStat(
      a.max_mem_usage,
      b.max_mem_usage,
      countA,
      countB,
    ),
    contention_time: addMaybeUnsetNumericStat(
      a.contention_time,
      b.contention_time,
      countA,
      countB,
    ),
    network_messages: addMaybeUnsetNumericStat(
      a.network_messages,
      b.network_messages,
      countA,
      countB,
    ),
    max_disk_usage: addMaybeUnsetNumericStat(
      a.max_disk_usage,
      b.max_disk_usage,
      countA,
      countB,
    ),
    cpu_sql_nanos: addMaybeUnsetNumericStat(
      a.cpu_sql_nanos,
      b.cpu_sql_nanos,
      countA,
      countB,
    ),
  };
}

export function addStatementStats(
  a: StatementStatistics,
  b: StatementStatistics,
): Required<StatementStatistics> {
  const countA = FixLong(a.count).toInt();
  const countB = FixLong(b.count).toInt();

  let regions: string[] = [];
  if (a.regions && b.regions) {
    regions = unique(a.regions.concat(b.regions));
  } else if (a.regions) {
    regions = a.regions;
  } else if (b.regions) {
    regions = b.regions;
  }

  let planGists: string[] = [];
  if (a.plan_gists && b.plan_gists) {
    planGists = unique(a.plan_gists.concat(b.plan_gists));
  } else if (a.plan_gists) {
    planGists = a.plan_gists;
  } else if (b.plan_gists) {
    planGists = b.plan_gists;
  }

  let indexRec: string[] = [];
  if (a.index_recommendations && b.index_recommendations) {
    indexRec = unique(a.index_recommendations.concat(b.index_recommendations));
  } else if (a.index_recommendations) {
    indexRec = a.index_recommendations;
  } else if (b.index_recommendations) {
    indexRec = b.index_recommendations;
  }

  let indexes: string[] = [];
  if (a.indexes && b.indexes) {
    indexes = unique(a.indexes.concat(b.indexes));
  } else if (a.indexes) {
    indexes = a.indexes;
  } else if (b.indexes) {
    indexes = b.indexes;
  }

  return {
    count: a.count.add(b.count),
    failure_count: a.failure_count.add(b.failure_count),
    first_attempt_count: a.first_attempt_count.add(b.first_attempt_count),
    max_retries: a.max_retries.greaterThan(b.max_retries)
      ? a.max_retries
      : b.max_retries,
    num_rows: aggregateNumericStats(a.num_rows, b.num_rows, countA, countB),
    idle_lat: aggregateNumericStats(a.idle_lat, b.idle_lat, countA, countB),
    parse_lat: aggregateNumericStats(a.parse_lat, b.parse_lat, countA, countB),
    plan_lat: aggregateNumericStats(a.plan_lat, b.plan_lat, countA, countB),
    run_lat: aggregateNumericStats(a.run_lat, b.run_lat, countA, countB),
    service_lat: aggregateNumericStats(
      a.service_lat,
      b.service_lat,
      countA,
      countB,
    ),
    overhead_lat: aggregateNumericStats(
      a.overhead_lat,
      b.overhead_lat,
      countA,
      countB,
    ),
    bytes_read: aggregateNumericStats(
      a.bytes_read,
      b.bytes_read,
      countA,
      countB,
    ),
    rows_read: aggregateNumericStats(a.rows_read, b.rows_read, countA, countB),
    rows_written: aggregateNumericStats(
      a.rows_written,
      b.rows_written,
      countA,
      countB,
    ),
    sensitive_info: coalesceSensitiveInfo(a.sensitive_info, b.sensitive_info),
    legacy_last_err: "",
    legacy_last_err_redacted: "",
    exec_stats: addExecStats(a.exec_stats, b.exec_stats),
    sql_type: a.sql_type,
    last_exec_timestamp:
      a.last_exec_timestamp &&
      b.last_exec_timestamp &&
      a.last_exec_timestamp.seconds > b.last_exec_timestamp.seconds
        ? a.last_exec_timestamp
        : b.last_exec_timestamp,
    nodes: uniqueLong([...a.nodes, ...b.nodes]),
    kv_node_ids: unique([...a.kv_node_ids, ...b.kv_node_ids]),
    regions: regions,
    used_follower_read: a.used_follower_read || b.used_follower_read,
    plan_gists: planGists,
    index_recommendations: indexRec,
    indexes: indexes,
    latency_info: aggregateLatencyInfo(a, b),
    last_error_code: "",
  };
}

export interface ExecutionStatistics {
  statement_fingerprint_id: Long;
  statement: string;
  statement_summary: string;
  aggregated_ts: number;
  aggregation_interval: number;
  app: string;
  database: string;
  distSQL: boolean;
  vec: boolean;
  implicit_txn: boolean;
  full_scan: boolean;
  node_id: number;
  txn_fingerprint_ids: Long[];
  stats: StatementStatistics;
}

export function flattenStatementStats(
  statementStats: CollectedStatementStatistics[],
): ExecutionStatistics[] {
  return statementStats.map(stmt => ({
    statement_fingerprint_id: stmt.id,
    statement: stmt.key.key_data.query,
    statement_summary: stmt.key.key_data.query_summary,
    aggregated_ts: TimestampToNumber(stmt.key.aggregated_ts),
    aggregation_interval: DurationToNumber(stmt.key.aggregation_interval),
    app: stmt.key.key_data.app,
    database: stmt.key.key_data.database,
    distSQL: stmt.key.key_data.distSQL,
    vec: stmt.key.key_data.vec,
    implicit_txn: stmt.key.key_data.implicit_txn,
    full_scan: stmt.key.key_data.full_scan,
    node_id: stmt.key.node_id,
    txn_fingerprint_ids: stmt.txn_fingerprint_ids,
    stats: stmt.stats,
  }));
}

// This function returns a key based on all parameters
// that should be used to group statements.
// Currently, using only statement_fingerprint_id
// (created by ConstructStatementFingerprintID using:
// query, implicit_txn, database, failed).
export function statementKey(stmt: ExecutionStatistics): string {
  return stmt.statement_fingerprint_id?.toString();
}

// transactionScopedStatementKey is similar to statementKey, except that
// it appends the transactionFingerprintID to the string key it generated.
export function transactionScopedStatementKey(
  stmt: ExecutionStatistics,
): string {
  return statementKey(stmt) + stmt.txn_fingerprint_ids?.toString() + stmt.app;
}

export const generateStmtDetailsToID = (
  fingerprintID: string,
  appNames: string,
  start: Long,
  end: Long,
): string => {
  if (
    appNames &&
    (appNames.includes("$ internal") || appNames.includes("unset"))
  ) {
    const apps = appNames.split(",");
    for (let i = 0; i < apps.length; i++) {
      if (apps[i].includes("$ internal")) {
        apps[i] = "$ internal";
      }
      if (apps[i].includes("unset")) {
        apps[i] = "";
      }
    }
    appNames = unique(apps).sort().toString();
  }
  let generatedID = fingerprintID;
  if (appNames) {
    generatedID += `/${appNames}`;
  }
  if (start) {
    generatedID += `/${start}`;
  }
  if (end) {
    generatedID += `/${end}`;
  }
  return generatedID;
};

export const generateTableID = (db: string, table: string): string => {
  return `${encodeURIComponent(db)}/${encodeURIComponent(table)}`;
};

File: pkg/sql/appstatspb/app_stats.go
// Copyright 2017 The Cockroach Authors.
//
// Use of this software is governed by the CockroachDB Software License
// included in the /LICENSE file.

package appstatspb

import (
	"math"
	"strconv"

	"github.com/cockroachdb/cockroach/pkg/util"
)

// StmtFingerprintID is the type of a Statement's fingerprint ID.
type StmtFingerprintID uint64

func (s StmtFingerprintID) String() string {
	return strconv.FormatUint(uint64(s), 10)
}

// ConstructStatementFingerprintID constructs an ID by hashing query with
// constants redacted, its database, and if it was part of an
// implicit txn. At the time of writing, these are the axis' we use to bucket
// queries for stats collection (see stmtKey).
var ConstructStatementFingerprintID = func(
	stmtNoConstants string, implicitTxn bool, database string,
) StmtFingerprintID {
	fnv := util.MakeFNV64()
	for _, c := range stmtNoConstants {
		fnv.Add(uint64(c))
	}
	for _, c := range database {
		fnv.Add(uint64(c))
	}
	if implicitTxn {
		fnv.Add('I')
	} else {
		fnv.Add('E')
	}
	return StmtFingerprintID(fnv.Sum())
}

// TransactionFingerprintID is the hashed string constructed using the
// individual statement fingerprint IDs that comprise the transaction.
type TransactionFingerprintID uint64

func (t TransactionFingerprintID) String() string {
	return strconv.FormatUint(uint64(t), 10)
}

// InvalidTransactionFingerprintID denotes an invalid transaction fingerprint ID.
const InvalidTransactionFingerprintID = TransactionFingerprintID(0)

// Size returns the size of the TransactionFingerprintID.
func (t TransactionFingerprintID) Size() int64 {
	return 8
}

// GetVariance retrieves the variance of the values.
func (l *NumericStat) GetVariance(count int64) float64 {
	return l.SquaredDiffs / (float64(count) - 1)
}

// Record updates the underlying running counts, incorporating the given value.
// It follows Welford's algorithm (Technometrics, 1962). The running count must
// be stored as it is required to finalize and retrieve the variance.
func (l *NumericStat) Record(count int64, val float64) {
	delta := val - l.Mean
	l.Mean += delta / float64(count)
	l.SquaredDiffs += delta * (val - l.Mean)
}

// Add combines b into this derived statistics.
func (l *NumericStat) Add(b NumericStat, countA, countB int64) {
	*l = AddNumericStats(*l, b, countA, countB)
}

// AlmostEqual compares two NumericStats between a window of size eps.
func (l *NumericStat) AlmostEqual(b NumericStat, eps float64) bool {
	return math.Abs(l.Mean-b.Mean) <= eps &&
		math.Abs(l.SquaredDiffs-b.SquaredDiffs) <= eps
}

// AddNumericStats combines derived statistics.
// Adapted from https://www.johndcook.com/blog/skewness_kurtosis/
func AddNumericStats(a, b NumericStat, countA, countB int64) NumericStat {
	total := float64(countA + countB)
	delta := b.Mean - a.Mean

	return NumericStat{
		Mean: ((a.Mean * float64(countA)) + (b.Mean * float64(countB))) / total,
		SquaredDiffs: (a.SquaredDiffs + b.SquaredDiffs) +
			delta*delta*float64(countA)*float64(countB)/total,
	}
}

// GetScrubbedCopy returns a copy of the given SensitiveInfo with its fields redacted
// or omitted entirely. By default, fields are omitted: if a new field is
// added to the SensitiveInfo proto, it must be added here to make it to the
// reg cluster.
func (si SensitiveInfo) GetScrubbedCopy() SensitiveInfo {
	output := SensitiveInfo{}
	// TODO(knz): This should really use si.LastErrorRedacted, however
	// this does not exist yet.
	// See: https://github.com/cockroachdb/cockroach/issues/53191
	output.LastErr = "<redacted>"
	// Not copying over MostRecentPlanDescription until we have an algorithm to scrub plan nodes.
	return output
}

// Add combines other into this TxnStats.
func (s *TxnStats) Add(other TxnStats) {
	s.TxnTimeSec.Add(other.TxnTimeSec, s.TxnCount, other.TxnCount)
	s.TxnCount += other.TxnCount
	s.ImplicitCount += other.ImplicitCount
	s.CommittedCount += other.CommittedCount
}

// Add combines other into TransactionStatistics.
func (t *TransactionStatistics) Add(other *TransactionStatistics) {
	if other.MaxRetries > t.MaxRetries {
		t.MaxRetries = other.MaxRetries
	}

	t.IdleLat.Add(other.IdleLat, t.Count, other.Count)
	t.CommitLat.Add(other.CommitLat, t.Count, other.Count)
	t.RetryLat.Add(other.RetryLat, t.Count, other.Count)
	t.ServiceLat.Add(other.ServiceLat, t.Count, other.Count)
	t.NumRows.Add(other.NumRows, t.Count, other.Count)
	t.RowsRead.Add(other.RowsRead, t.Count, other.Count)
	t.BytesRead.Add(other.BytesRead, t.Count, other.Count)
	t.RowsWritten.Add(other.RowsWritten, t.Count, other.Count)

	t.ExecStats.Add(other.ExecStats)

	t.Count += other.Count
}

// Add combines CollectedStatementStatistics into a single AggregatedStatementMetadata.
func (s *AggregatedStatementMetadata) Add(other *CollectedStatementStatistics) {
	// Only set the value if it hasn't already been set.
	if s.Query == "" || s.QuerySummary == "" {
		s.ImplicitTxn = other.Key.ImplicitTxn
		s.Query = other.Key.Query
		s.QuerySummary = other.Key.QuerySummary
		s.StmtType = other.Stats.SQLType
	}

	// Avoid creating the array if the db names match.
	if len(s.Databases) != 1 || s.Databases[0] != other.Key.Database {
		s.Databases = util.CombineUnique(s.Databases, []string{other.Key.Database})
	}

	if other.Key.DistSQL {
		s.DistSQLCount++
	}
	if other.Key.FullScan {
		s.FullScanCount++
	}
	if other.Key.Vec {
		s.VecCount++
	}
	s.TotalCount++
}

// Add combines other into this StatementStatistics.
func (s *StatementStatistics) Add(other *StatementStatistics) {
	s.FirstAttemptCount += other.FirstAttemptCount
	if other.MaxRetries > s.MaxRetries {
		s.MaxRetries = other.MaxRetries
	}
	s.SQLType = other.SQLType
	s.NumRows.Add(other.NumRows, s.Count, other.Count)
	s.IdleLat.Add(other.IdleLat, s.Count, other.Count)
	s.ParseLat.Add(other.ParseLat, s.Count, other.Count)
	s.PlanLat.Add(other.PlanLat, s.Count, other.Count)
	s.RunLat.Add(other.RunLat, s.Count, other.Count)
	s.ServiceLat.Add(other.ServiceLat, s.Count, other.Count)
	s.OverheadLat.Add(other.OverheadLat, s.Count, other.Count)
	s.BytesRead.Add(other.BytesRead, s.Count, other.Count)
	s.RowsRead.Add(other.RowsRead, s.Count, other.Count)
	s.RowsWritten.Add(other.RowsWritten, s.Count, other.Count)
	s.Nodes = util.CombineUnique(s.Nodes, other.Nodes)
	s.KVNodeIDs = util.CombineUnique(s.KVNodeIDs, other.KVNodeIDs)
	s.Regions = util.CombineUnique(s.Regions, other.Regions)
	s.UsedFollowerRead = s.UsedFollowerRead || other.UsedFollowerRead
	s.PlanGists = util.CombineUnique(s.PlanGists, other.PlanGists)
	s.Indexes = util.CombineUnique(s.Indexes, other.Indexes)
	s.ExecStats.Add(other.ExecStats)
	s.LatencyInfo.MergeMaxMin(other.LatencyInfo)

	if s.SensitiveInfo.MostRecentPlanTimestamp.Before(other.SensitiveInfo.MostRecentPlanTimestamp) {
		s.SensitiveInfo = other.SensitiveInfo
	}

	// Use the LastExecTimestamp to decide which object has the last error.
	if s.LastExecTimestamp.Before(other.LastExecTimestamp) {
		s.LastExecTimestamp = other.LastExecTimestamp

		if other.SensitiveInfo.LastErr != "" {
			s.SensitiveInfo.LastErr = other.SensitiveInfo.LastErr
		}

		if other.LastErrorCode != "" {
			s.LastErrorCode = other.LastErrorCode
		}
	}

	if len(other.IndexRecommendations) > 0 {
		s.IndexRecommendations = other.IndexRecommendations
	}

	s.Count += other.Count
	s.FailureCount += other.FailureCount
}

// AlmostEqual compares two StatementStatistics and their contained NumericStats
// objects within an window of size eps, ExecStats are ignored.
func (s *StatementStatistics) AlmostEqual(other *StatementStatistics, eps float64) bool {
	return s.Count == other.Count &&
		s.FirstAttemptCount == other.FirstAttemptCount &&
		s.MaxRetries == other.MaxRetries &&
		s.NumRows.AlmostEqual(other.NumRows, eps) &&
		s.IdleLat.AlmostEqual(other.IdleLat, eps) &&
		s.ParseLat.AlmostEqual(other.ParseLat, eps) &&
		s.PlanLat.AlmostEqual(other.PlanLat, eps) &&
		s.RunLat.AlmostEqual(other.RunLat, eps) &&
		s.ServiceLat.AlmostEqual(other.ServiceLat, eps) &&
		s.OverheadLat.AlmostEqual(other.OverheadLat, eps) &&
		s.SensitiveInfo.Equal(other.SensitiveInfo) &&
		s.BytesRead.AlmostEqual(other.BytesRead, eps) &&
		s.RowsRead.AlmostEqual(other.RowsRead, eps) &&
		s.RowsWritten.AlmostEqual(other.RowsWritten, eps)
	// s.ExecStats are deliberately ignored since they are subject to sampling
	// probability and are not fully deterministic (e.g. the number of network
	// messages depends on the range cache state).
}

// Add combines other into this ExecStats.
func (s *ExecStats) Add(other ExecStats) {
	// Execution stats collected using a sampling approach.
	execStatCollectionCount := s.Count
	if execStatCollectionCount == 0 && other.Count == 0 {
		// If both are zero, artificially set the receiver's count to one to avoid
		// division by zero in Add.
		execStatCollectionCount = 1
	}
	s.NetworkBytes.Add(other.NetworkBytes, execStatCollectionCount, other.Count)
	s.MaxMemUsage.Add(other.MaxMemUsage, execStatCollectionCount, other.Count)
	s.ContentionTime.Add(other.ContentionTime, execStatCollectionCount, other.Count)
	s.NetworkMessages.Add(other.NetworkMessages, execStatCollectionCount, other.Count)
	s.MaxDiskUsage.Add(other.MaxDiskUsage, execStatCollectionCount, other.Count)
	s.CPUSQLNanos.Add(other.CPUSQLNanos, execStatCollectionCount, other.Count)

	s.MVCCIteratorStats.StepCount.Add(other.MVCCIteratorStats.StepCount, execStatCollectionCount, other.Count)
	s.MVCCIteratorStats.StepCountInternal.Add(other.MVCCIteratorStats.StepCountInternal, execStatCollectionCount, other.Count)
	s.MVCCIteratorStats.SeekCount.Add(other.MVCCIteratorStats.SeekCount, execStatCollectionCount, other.Count)
	s.MVCCIteratorStats.SeekCountInternal.Add(other.MVCCIteratorStats.SeekCountInternal, execStatCollectionCount, other.Count)
	s.MVCCIteratorStats.BlockBytes.Add(other.MVCCIteratorStats.BlockBytes, execStatCollectionCount, other.Count)
	s.MVCCIteratorStats.BlockBytesInCache.Add(other.MVCCIteratorStats.BlockBytesInCache, execStatCollectionCount, other.Count)
	s.MVCCIteratorStats.KeyBytes.Add(other.MVCCIteratorStats.KeyBytes, execStatCollectionCount, other.Count)
	s.MVCCIteratorStats.ValueBytes.Add(other.MVCCIteratorStats.ValueBytes, execStatCollectionCount, other.Count)
	s.MVCCIteratorStats.PointCount.Add(other.MVCCIteratorStats.PointCount, execStatCollectionCount, other.Count)
	s.MVCCIteratorStats.PointsCoveredByRangeTombstones.Add(other.MVCCIteratorStats.PointsCoveredByRangeTombstones, execStatCollectionCount, other.Count)
	s.MVCCIteratorStats.RangeKeyCount.Add(other.MVCCIteratorStats.RangeKeyCount, execStatCollectionCount, other.Count)
	s.MVCCIteratorStats.RangeKeyContainedPoints.Add(other.MVCCIteratorStats.RangeKeyContainedPoints, execStatCollectionCount, other.Count)
	s.MVCCIteratorStats.RangeKeySkippedPoints.Add(other.MVCCIteratorStats.RangeKeySkippedPoints, execStatCollectionCount, other.Count)

	s.Count += other.Count
}

// MergeMaxMin combines the max and min only into this LatencyInfo.
func (s *LatencyInfo) MergeMaxMin(other LatencyInfo) {
	if s.Min == 0 || other.Min < s.Min {
		s.Min = other.Min
	}
	if other.Max > s.Max {
		s.Max = other.Max
	}
}
